### ISSUE TITLE
New Lint: ambiguous closures like `return || true`

### ISSUE THREAD

**Issue by yaahc** [rust-lang/rust-clippy#4374]
The proposal is to implement a new lint to capture closures that are appear ambiguous to the reader. Such closures have no args, which makes the leading `||` look like a logical OR, and have an expression body without surrounding block. It may also only be ambiguous if the closure's body expression returns a boolean.

the classic example is `return || true` which at first glance looks like nonsense but actually returns a closure that always returns true.

Clippy should warn that this closure is potentially ambiguous to the reader and may be misinterpreted as boolean logic. It would also be nice if it suggests a machine applicable solution so that rustfix can benefit from the lint.

Possible suggestions include

```rust
return || { true }

return { || true }
```

I'm partial to the former.

**Comment by llogiq** [comment]
Why choose a block over parentheses?

**Comment by llogiq** [comment]
The reason I'm asking is that a block, unlike parentheses, introduces a new scope.

**Comment by yaahc** [comment]
There was no reason for not also suggesting parentheses, just an oversight. Though some quick testing shows that a lint currently exists in rustc to warn against unnecessary parenthesis.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5a7f977cd95e0f3b7c0968b7efe7c416

where as a similar lint does not exist for the surrounding block version.

**Comment by llogiq** [comment]
I don't see a problem with changing the rustc lint to no longer trigger in that particular case, but if we choose the block variant, we better make sure that we don't run afoul of borrowck.
